testgen/src/tester.rs

```diff
@@ -0,0 +1,337 @@
+use crate::helpers::*;
+use crate::tester::TestResult::{Failure, ParseError, ReadError, Success};
+use serde::de::DeserializeOwned;
+use std::panic::UnwindSafe;
+use std::{fs, path::PathBuf};
+use std::{
+    panic,
+    sync::{Arc, Mutex},
+};
+
+#[derive(Debug, Clone)]
+pub struct TestEnv {
+    root_dir: String,
+    logs: Vec<String>,
+}
+
+impl TestEnv {
+    pub fn add_log(&mut self, log: &str) {
+        self.logs.push(log.to_string());
+    }
+
+    /// Read a file from a path relative to the environment root into a string
+    pub fn read_file(&mut self, path: &str) -> Option<String> {
+        match self.full_path(path) {
+            None => None,
+            Some(full_path) => match fs::read_to_string(&full_path) {
+                Ok(file) => Some(file),
+                Err(_) => None,
+            },
+        }
+    }
+
+    /// Convert a relative path to the full path from the test root
+    /// Return None the full path can't be formed
+    pub fn full_path(&self, rel_path: &str) -> Option<String> {
+        let full_path = PathBuf::from(&self.root_dir).join(rel_path);
+        match full_path.to_str() {
+            None => None,
+            Some(full_path) => Some(full_path.to_string()),
+        }
+    }
+
+    /// Convert a full path to the path relative to the test root
+    /// Return None if the full path doesn't contain test root as prefix
+    pub fn rel_path(&self, full_path: &str) -> Option<String> {
+        match PathBuf::from(full_path).strip_prefix(&self.root_dir) {
+            Err(_) => None,
+            Ok(rel_path) => match rel_path.to_str() {
+                None => None,
+                Some(rel_path) => Some(rel_path.to_string()),
+            },
+        }
+    }
+}
+
+pub struct Tester {
+    root_dir: String,
+    tests: Vec<(String, Box<dyn Fn(&str) -> TestResult>)>,
```

Comment:
There's a clippy warning here for type complexity. I'm OKAY with this one being complex because it's not that difficult to read here but i'm not sure how much we adhere to clippy warnings

Reply:
Oh, the warning is about this being complex

Reply:
done; the types will also change in further PRs